### PR TITLE
Hardening flags

### PR DESCRIPTION
**Pull request by RossBrunton**
UR_DEVELOPER_MODE has also been changed to only set Werror. The
previous other flags set by this variable either have no effect
(-fno-omit-frame-pointer) or are now always enabled
(-fstack-protector-strong).

Some smaller warnings have also been fixed.

### CODE REVIEW

File: cmake/helpers.cmake [hunk 1]
```diff
@@ -59,17 +59,32 @@ endmacro()
...
+                -fcf-protection=full
+                -fstack-protector-strong
+                -fvisibility=hidden # Required for -fsanitize=cfi
+                # -fsanitize=cfi requires -flto, which breaks a lot of things
+                # See: https://github.com/oneapi-src/unified-runtime/issues/2120
+                # -flto
+                # $<$<CXX_COMPILER_ID:Clang,AppleClang>:-fsanitize=cfi>
+            )
+            if (CMAKE_BUILD_TYPE STREQUAL "Release")
+                target_compile_options(${name} PRIVATE -Werror)
```
Comment (aarongreig) [inline cmake/helpers.cmake:84]: is this just for convenience if you're hacking around in debug? I'd vote for enabling it all the time if we can

File: source/adapters/level_zero/CMakeLists.txt [hunk 2]
```diff
@@ -66,6 +66,21 @@ if(UR_BUILD_ADAPTER_L0)
         )
     endif()
 
+    # Ensure UR flags are propogated to level zero
```
Comment (aarongreig) [inline source/adapters/level_zero/CMakeLists.txt:69]:
```suggestion
    # Ensure UR flags are propagated to level zero
```

**Review by kbenzie — APPROVED** [review]

**Comment by RossBrunton** [comment]
@oneapi-src/unified-runtime-level-zero-write Are you okay with this change? It's adding a bunch of "hardening" flags to the loader targets.

**Comment by pbalcer** [comment]
> @oneapi-src/unified-runtime-level-zero-write Are you okay with this change? It's adding a bunch of "hardening" flags to the loader targets.

We can do that, but it will only affect UR and intel/llvm builds. Normally the loader is shipped from the driver package.

My (very slight) preference would be that we don't set any flags that might affect the performance or cause difference in behavior compared to the unmodified loader.

**Review by pbalcer — APPROVED** [review]

**Pull request merged**